moveit_core/robot_state/src/robot_state.cpp

```diff
@@ -1945,23 +1945,75 @@ double RobotState::computeCartesianPath(const JointModelGroup* group, std::vecto
   }
 
-  if (test_joint_space_jump)
+  last_valid_percentage *= testJointSpaceJump(group, traj, jump_threshold);
+
+  return last_valid_percentage;
+}
+
+double RobotState::computeCartesianPath(const JointModelGroup* group, std::vector<RobotStatePtr>& traj,
+                                        const LinkModel* link, const EigenSTL::vector_Affine3d& waypoints,
+                                        bool global_reference_frame, double max_step,
+                                        const JumpThreshold& jump_threshold,
+                                        const GroupStateValidityCallbackFn& validCallback,
+                                        const kinematics::KinematicsQueryOptions& options)
+{
+  double percentage_solved = 0.0;
+  for (std::size_t i = 0; i < waypoints.size(); ++i)
   {
-    last_valid_percentage *= testJointSpaceJump(group, traj, jump_threshold);
+    // Don't test joint space jumps for every waypoint, test them later on the whole trajectory.
+    static const double no_joint_space_jump_test = 0.0;
+    std::vector<RobotStatePtr> waypoint_traj;
+    double wp_percentage_solved = computeCartesianPath(group, waypoint_traj, link, waypoints[i], global_reference_frame,
+                                                       max_step, no_joint_space_jump_test, validCallback, options);
+    if (fabs(wp_percentage_solved - 1.0) < std::numeric_limits<double>::epsilon())
+    {
+      percentage_solved = (double)(i + 1) / (double)waypoints.size();
+      std::vector<RobotStatePtr>::iterator start = waypoint_traj.begin();
+      if (i > 0 && !waypoint_traj.empty())
+        std::advance(start, 1);
+      traj.insert(traj.end(), start, waypoint_traj.end());
+    }
+    else
+    {
+      percentage_solved += wp_percentage_solved / (double)waypoints.size();
+      std::vector<RobotStatePtr>::iterator start = waypoint_traj.begin();
+      if (i > 0 && !waypoint_traj.empty())
+        std::advance(start, 1);
+      traj.insert(traj.end(), start, waypoint_traj.end());
+      break;
+    }
   }
 
-  return last_valid_percentage;
+  percentage_solved *= testJointSpaceJump(group, traj, jump_threshold);
+
+  return percentage_solved;
+}
+
+double RobotState::testJointSpaceJump(const JointModelGroup* group, std::vector<RobotStatePtr>& traj,
+                                      const JumpThreshold& jt)
+{
+  if (jt.jump_threshold_factor > 0.0)
+    return testJointSpaceJump(group, traj, jt.jump_threshold_factor);
+
+  else if (jt.prismatic_jump_threshold > 0.0 && jt.revolute_jump_threshold > 0.0)
+    return testJointSpaceJump(group, traj, jt.revolute_jump_threshold, jt.prismatic_jump_threshold);
+
+  else if (jt.prismatic_jump_threshold > 0.0 || jt.revolute_jump_threshold > 0.0)
+    CONSOLE_BRIDGE_logWarn("Attempting to test for absolute joint-space jumps, but zero threshold was provided");
+
+  return 1.0;
 }
 
 double RobotState::testJointSpaceJump(const JointModelGroup* group, std::vector<RobotStatePtr>& traj,
```

Comment:
What about renaming these worker functions to `testJointSpaceJumpRelative` and `testJointSpaceJumpAbsolute`?

Reply:
I think that's a good idea